pkg/controller/certificate-shim/sync.go

```diff
@@ -0,0 +1,598 @@
+	if hasIngressClassVal {
+		if crt.Annotations == nil {
+			crt.Annotations = make(map[string]string)
+		}
+		crt.Annotations[cmacme.ACMECertificateHTTP01IngressClassOverride] = ingressClassVal
+	}
+
+	ingLike.SetAnnotations(ingAnnotations)
+}
+
+// hasShimAnnotation returns true if this ingress-like object contains one of
+// the annotations "cert-manager.io/issuer", "cert-manager.io/cluster-issuer",
+// or one of the annotations provided with --auto-certificate-annotations (which
+// default to "kubernetes.io/tls-acme").
+func hasShimAnnotation(ingLike metav1.Object, autoCertificateAnnotations []string) bool {
+	annotations := ingLike.GetAnnotations()
+	if annotations == nil {
+		annotations = map[string]string{}
+	}
+	if _, ok := annotations[cmapi.IngressIssuerNameAnnotationKey]; ok {
+		return true
+	}
+	if _, ok := annotations[cmapi.IngressClusterIssuerNameAnnotationKey]; ok {
+		return true
+	}
+	for _, x := range autoCertificateAnnotations {
+		if s, ok := annotations[x]; ok {
+			if b, _ := strconv.ParseBool(s); b {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// issuerForIngressLike determines the Issuer that should be specified on a
+// Certificate created for the given ingress-like resource. If one is not set,
+// the default issuer given to the controller is used.
+func issuerForIngressLike(defaults controller.IngressShimOptions, ingLike metav1.Object) (name, kind, group string, err error) {
+	var errs []string
+
+	name = defaults.DefaultIssuerName
+	kind = defaults.DefaultIssuerKind
+	group = defaults.DefaultIssuerGroup
+
+	annotations := ingLike.GetAnnotations()
+
+	if annotations == nil {
+		annotations = map[string]string{}
+	}
+
+	issuerName, issuerNameOK := annotations[cmapi.IngressIssuerNameAnnotationKey]
+	if issuerNameOK {
+		name = issuerName
+		kind = cmapi.IssuerKind
+	}
+
+	clusterIssuerName, clusterIssuerNameOK := annotations[cmapi.IngressClusterIssuerNameAnnotationKey]
+	if clusterIssuerNameOK {
+		name = clusterIssuerName
+		kind = cmapi.ClusterIssuerKind
+	}
+
+	kindName, kindNameOK := annotations[cmapi.IssuerKindAnnotationKey]
+	if kindNameOK {
+		kind = kindName
+	}
+
+	groupName, groupNameOK := annotations[cmapi.IssuerGroupAnnotationKey]
+	if groupNameOK {
+		group = groupName
+	}
+
+	if len(name) == 0 {
+		errs = append(errs, "failed to determine issuer name to be used for ingress resource")
+	}
+
+	if issuerNameOK && clusterIssuerNameOK {
+		errs = append(errs,
+			fmt.Sprintf("both %q and %q may not be set",
+				cmapi.IngressIssuerNameAnnotationKey, cmapi.IngressClusterIssuerNameAnnotationKey))
+	}
+
+	if clusterIssuerNameOK && groupNameOK {
+		errs = append(errs,
+			fmt.Sprintf("both %q and %q may not be set",
+				cmapi.IngressClusterIssuerNameAnnotationKey, cmapi.IssuerGroupAnnotationKey))
+	}
+
+	if clusterIssuerNameOK && kindNameOK {
+		errs = append(errs,
+			fmt.Sprintf("both %q and %q may not be set",
+				cmapi.IngressClusterIssuerNameAnnotationKey, cmapi.IssuerKindAnnotationKey))
+	}
```

Comment:
Question: is it not a valid scenario to pass a cluster issuer with a non-default kind/group? For example Google CAS issuer has a cluster issuer type https://github.com/jetstack/google-cas-issuer#configuring-the-issuer if one wanted to specify that the cert needs to be provided by this a Google CAS cluster issuer, what annotations would they use? Or is it the case that only ACME issuer is ever used here.

Reply:
I agree, I find the above `clusterIssuerNameOK && kindNameOK` surprising. These two `if` statements have been added in 2019 in https://github.com/jetstack/cert-manager/pull/2105.

I would expect this to work:

```yaml
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  annotations:
    cert-manager.io/cluster-issuer: my-google-cas-issuer
    cert-manager.io/issuer-group: cas-issuer.jetstack.io
    cert-manager.io/issuer-kind: GoogleCASClusterIssuer
```

Is there something that is preventing us from doing this? @JoshVanL



```yaml
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  annotations:
    cert-manager.io/issuer-group: cas-issuer.jetstack.io
    cert-manager.io/issuer-kind: GoogleCASClusterIssuer
```

Reply:
Since it's the already existing behaviour, I think it's fine for this PR